demonstrations/tutorial_local_cost_functions.py

```diff
@@ -0,0 +1,432 @@
+# :math:`\theta`\ 's for each rotation
+#
+
+RX = np.random.uniform(low=-np.pi, high=np.pi)
+RY = np.random.uniform(low=-np.pi, high=np.pi)
+rotations = [[RX for i in range(wires)], [RY for i in range(wires)]]
+
+
+######################################################################
+# And look at the results:
+#
+
+print("Global Cost: {: .7f}".format(cost_global(rotations)))
+print("Local Cost: {: .7f}".format(cost_local(rotations)))
+print("--- Global Circuit ---")
+print(global_circuit.draw())
+print("--- Local Circuit")
+print(local_circuit.draw())
+
+
+######################################################################
+# With this simple example, we can visualize the cost function, and see
+# the barren plateau effect graphically.
+
+X = np.arange(-np.pi, np.pi, 1)
+Y = np.arange(-np.pi, np.pi, 1)
+X, Y = np.meshgrid(X, Y)
+
+local_Z = []
+global_Z = []
+local_z = []
+global_z = []
+
+X = np.arange(-np.pi, np.pi, 0.25)
+Y = np.arange(-np.pi, np.pi, 0.25)
+X, Y = np.meshgrid(X, Y)
+
+for x in X[0, :]:
+    for y in Y[:, 0]:
+        rotations = [[x for i in range(wires)], [y for i in range(wires)]]
+        local_z.append(cost_local(rotations))
+        global_z.append(cost_global(rotations))
+    local_Z.append(local_z)
+    global_Z.append(global_z)
+    local_z = []
+    global_z = []
+
+local_Z = np.asarray(local_Z)
+global_Z = np.asarray(global_Z)
+
+
+######################################################################
+# Firstly, we look at the global cost function. When plotting the cost
+# function across 6 qubits, much of the cost landscape is flat, and
+# difficult to train (even with a circuit depth of only 2!). This effect
+# will worsen as the number of qubits increases.
+#
+
+fig = plt.figure()
+ax = fig.add_subplot(111, projection="3d")
+surf = ax.plot_surface(X, Y, global_Z, cmap="viridis", linewidth=0, antialiased=False)
+ax.set_zlim(0, 1)
+ax.zaxis.set_major_locator(LinearLocator(10))
+ax.zaxis.set_major_formatter(FormatStrFormatter("%.02f"))
+plt.show()
+
+
+######################################################################
+# However, when we change to the local cost function, the cost lanscape
+# becomes much more trainable.
+#
+
+fig = plt.figure()
+ax = fig.add_subplot(111, projection="3d")
+surf = ax.plot_surface(X, Y, local_Z, cmap="viridis", linewidth=0, antialiased=False)
+ax.set_zlim(0, 1)
+ax.zaxis.set_major_locator(LinearLocator(10))
+ax.zaxis.set_major_formatter(FormatStrFormatter("%.02f"))
+plt.show()
+
+
+######################################################################
+# Those are some nice pictures, but how do they refect actual
+# trainability? Lets try training both the local, and global cost
+# functions. Because we have a visualization of the total cost landscape,
+# lets pick a point to exaggerate the problem. The worst point in the
+# landscape is :math:`(\pi,0)`, so lets use that.
+#
+
+rotations = [[3 for i in range(wires)], [0 for i in range(wires)]]
```

Comment:
```suggestion
rotations = [[np.pi for i in range(wires)], [0 for i in range(wires)]]
```
If we put exactly on that point do things not train?

Reply:
Haha, no, unfortunately things do not train directly on the point. I didn't realize this and tore my hair out over it for a week.